lib/linked_list.rb

```diff
@@ -14,75 +14,171 @@ def initialize(value, next_node = nil)
 class LinkedList
     def initialize
       @head = nil # keep the head private. Not accessible outside this class
+      @tail = nil
     end
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(value)
-      raise NotImplementedError
+      if @head.nil?
+        @head = @tail = Node.new(value)
+      else
+        new_node = Node.new(value, @head)
+        @head = new_node
+      end

```

Comment:
This can be simplified somewhat. 

```suggestion
      @head = Node.new(value, @head)
```